javascript/pkce_oauth/server/index.js

```diff
@@ -0,0 +1,174 @@
+const express = require('express');
+const cookieParser = require('cookie-parser');
+const sha256 = require('js-sha256');
+const request = require('request');
+const functions = require('firebase-functions');
+
+const challengeMethod = "S256";
+const authorizationGrantType = "authorization_code";
+const refreshGrantType = "refresh_token";
+
+const clientId = process.env.client_id;
+const clientSecret = process.env.client_secret;
+const redirectUri = process.env.redirect_uri;
+const frontEndLocation = process.env.front_end_location;
+const tokenExchangeEndpoint = process.env.token_exchange_endpoint;
+
+const stateCache = {};
+const tokenCache = {};
+
+const app = express();
+app.use(cookieParser());
+
+// Base endpoint. This should return our html.
+app.get('/', (req, res) => {
+    console.log('In /');
+
+    request(frontEndLocation).pipe(res);
+});
+
+// First step.
+// We need to generate state and code verifiers for this client.
+// We then return the info the client needs to hit the auth server.
+app.get('/first-step-auth', (req, res) => {
+    console.log('In first-step-auth');
+
+    let result = {};
+
+    let state = getRandomToken();
+    let codeVerifier = getRandomToken();
+    let codeChallenge = sha256(codeVerifier);
+
+    // Save that the state and codeVerifier were given to the same person
+    stateCache[state] = codeVerifier;
+
+    // Supply the client with the OAuth data
+    result.codeChallenge = codeChallenge;
+    result.challengeMethod = challengeMethod;
+    result.clientId = clientId;
+    result.redirectUri = redirectUri;
+
+    // Store state as a cookie on the client
+    res.cookie("state", state);
+
+    res.json(result);
+});
+
+// Second step.
+// We need to get the state and code from the client.
+// We will verify the state is valid, and send the code to the auth server.
+// We then store the refresh token locally and return the access token to the client.
+app.get('/second-step-auth', (req, res) => {
+    console.log('In second-step-auth');
+
+    let state = req.query.state;
```

Comment:
```suggestion
    const state = req.query.state;
```

And in many other places too